File: AgoraChat-ObjectiveC/AgoraChat/Code/Controllers/Contacts/ACDContactListController.m
//
//  AgoraContactListController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/21.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDContactListController.h"
#import "MISScrollPage.h"
#import "AgoraContactListSectionHeader.h"
#import "ACDContactInfoViewController.h"

#import "AgoraChatroomsViewController.h"
#import "ACDContactCell.h"
#import "AgoraUserModel.h"
#import "AgoraApplyManager.h"
#import "AgoraChatDemoHelper.h"
#import "AgoraRealtimeSearchUtils.h"
#import "NSArray+AgoraSortContacts.h"
#import "ACDContactCell.h"
#import "PresenceManager.h"


@interface ACDContactListController()
@property (nonatomic,strong) NSArray<NSString*>* contacts;
@end

@implementation ACDContactListController

#pragma mark life cycle
- (instancetype)init {
    self = [super init];
    if (self) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(settingBlackListDidChange) name:@"AgoraSettingBlackListDidChange" object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(presencesUpdated:) name:PRESENCES_UPDATE object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(contactListDidChange) name:KACD_REFRESH_CONTACTS object:nil];
    }
    return  self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self useRefresh];
    
    [self tableDidTriggerHeaderRefresh];
}


#pragma mark refresh and load more
- (void)didStartRefresh {
    [self tableDidTriggerHeaderRefresh];
}

- (void)tableDidTriggerHeaderRefresh {
    if (self.isSearchState) {
        [self endRefresh];
        return;
    }
    
    ACD_WS
    [[AgoraChatClient sharedClient].contactManager getContactsFromServerWithCompletion:^(NSArray *aList, AgoraChatError *aError) {
        if (aError == nil) {
            [weakSelf updateContacts:aList];
            [weakSelf.table reloadData];
        }
        else {
//            [weakSelf tableViewDidFinishTriggerHeader:YES];
        }
        [weakSelf endRefresh];
    }];
}

- (void)loadContactsFromServer
{
    [self tableDidTriggerHeaderRefresh];
}

- (void)reloadContacts {
    NSArray *bubbyList = [[AgoraChatClient sharedClient].contactManager getContacts];
    [self updateContacts:bubbyList];
    WEAK_SELF
    dispatch_async(dispatch_get_main_queue(), ^(){
        [weakSelf.table reloadData];
    });
}


- (void)updateContacts:(NSArray *)bubbyList {
    self.contacts = bubbyList;
    NSArray *blockList = [[AgoraChatClient sharedClient].contactManager getBlackList];
    NSMutableArray *contacts = [NSMutableArray arrayWithArray:bubbyList];
    for (NSString *blockId in blockList) {
        [contacts removeObject:blockId];
    }
    [self sortContacts:contacts];
}

- (void)sortContacts:(NSArray *)contacts {
    if (contacts.count == 0) {
        self.dataArray = [@[] mutableCopy];
        self.sectionTitles = [@[] mutableCopy];
        self.searchSource = [@[] mutableCopy];
        return;
    }
    
    NSMutableArray *sectionTitles = nil;
    NSMutableArray *searchSource = nil;
    NSArray *sortArray = [NSArray sortContacts:contacts
                                 sectionTitles:&sectionTitles
                                  searchSource:&searchSource];
    [self.dataArray removeAllObjects];
    [self.dataArray addObjectsFromArray:sortArray];
    [[PresenceManager sharedInstance] subscribe:contacts completion:nil];
    self.sectionTitles = [NSMutableArray arrayWithArray:sectionTitles];
    self.searchSource = [NSMutableArray arrayWithArray:searchSource];
}

- (void)viewDidAppearForIndex:(NSUInteger)index{
    //[self reloadContacts];
}

#pragma mark NSNotification
- (void)contactListDidChange {
    [self reloadContacts];
}

#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (self.isSearchState) {
        return 1;
    }
    return  self.sectionTitles.count;
}

- (NSArray*)sectionIndexTitlesForTableView:(UITableView *)tableView{
     return self.sectionTitles;
}


- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index{
    return index;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if (self.isSearchState) {
        return self.searchResults.count;
    }
    return ((NSArray *)self.dataArray[section]).count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    ACDContactCell *cell = [tableView dequeueReusableCellWithIdentifier:[ACDContactCell reuseIdentifier]];
    if (cell == nil) {
        cell = [[ACDContactCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDContactCell reuseIdentifier]];
    }
    AgoraUserModel *model = nil;

    if (self.isSearchState) {
        model = self.searchResults[indexPath.row];
        cell.model = model;
    }else {
        model = self.dataArray[indexPath.section][indexPath.row];
        cell.model = model;
    }
    AgoraChatPresence*presence = [[PresenceManager sharedInstance].presences objectForKey:model.hyphenateId];
    if(presence) {
        NSInteger status = [PresenceManager fetchStatus:presence];
        NSString* imageName = [[PresenceManager whiteStrokePresenceImages] objectForKey:[NSNumber numberWithInteger:status]];
        [cell.iconImageView setPresenceImage:[UIImage imageNamed:imageName]];
        NSString* showStatus = [[PresenceManager showStatus] objectForKey:[NSNumber numberWithInteger:status]];
        if(status == 0)
            showStatus = [PresenceManager formatOfflineStatus:presence.lastTime];
        if(status != PRESENCESTATUS_OFFLINE && presence.statusDescription.length > 0)
            cell.detailLabel.text = presence.statusDescription;
        else
            cell.detailLabel.text = showStatus;
    }
    
    cell.tapCellBlock = ^{
        if (self.selectedBlock) {
            self.selectedBlock(model.hyphenateId);
        }
    };
    
    return cell;
}

#pragma mark getter and setter
- (UITableView *)table {
    if (_table == nil) {
        _table  = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, KScreenHeight) style:UITableViewStylePlain];
        _table.delegate        = self;
        _table.dataSource      = self;
        _table.separatorStyle  = UITableViewCellSeparatorStyleNone;
        _table.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
        _table.clipsToBounds = YES;
        _table.rowHeight = 54.0f;
        [_table registerClass:[ACDContactCell class] forCellReuseIdentifier:[ACDContactCell reuseIdentifier]];
        _table.sectionIndexColor = SectionIndexTextColor;
        _table.sectionIndexBackgroundColor = [UIColor clearColor];
        _table.allowsMultipleSelection = NO;

    }
    return _table;
}

- (void)presencesUpdated:(NSNotification*)noti
{
    NSArray*array = noti.object;
    NSMutableSet *set1 = [NSMutableSet setWithArray:array];
    NSMutableSet *set2 = [NSMutableSet setWithArray:self.contacts];
    [set1 intersectSet:set2];
    if(set1.count > 0) {
        __weak typeof(self) weakself = self;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakself.table reloadData];
        });
        
    }
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Contacts/ACDContactListController.swift
//
//  ACDContactListController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/17.
//

import UIKit

class ACDContactListController: ACDContainerSearchTableViewController<[AgoraUserModel]> {

    private var contacts: [String]?
    
    var selectedBlock: ((_ contactId: String) -> Void)?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        NotificationCenter.default.addObserver(self, selector: #selector(presencesUpdated(_:)), name: PresenceUpdateNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(contactListDidChange), name: ContactsUpdatedNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(didRecvUserInfoDidChangeNotification(_:)), name: UserInfoDidChangeNotification, object: nil)
        
        self.table.separatorStyle = .none
        self.table.keyboardDismissMode = .onDrag
        self.table.rowHeight = 54
        self.table.sectionIndexColor = Color_353535
        self.table.sectionIndexBackgroundColor = .clear
        self.table.register(UINib(nibName: "ACDContactCell", bundle: nil), forCellReuseIdentifier: "cell")
        
        self.useRefresh()
        self.didStartRefresh()
    }
    
    override func didStartRefresh() {
        if self.isSearchState {
            self.endRefresh()
            return
        }
        AgoraChatClient.shared().contactManager?.getContactsFromServer(completion: { list, error in
            if let list = list {
                UserInfoStore.shared.fetchUserInfosFromServer(userIds: list) {
                    self.table.reloadData()
                }
                self.updateContacts(bubbyList: list)
            }
            self.endRefresh()
        })
    }
    
    @objc private func presencesUpdated(_ notification: Notification) {
        guard let array = notification.object as? [String], let contacts = self.contacts else {
            return
        }
        for item in array where contacts.contains(item) {
            self.table.reloadData()
            break
        }
    }
    
    @objc private func contactListDidChange() {
        self.reloadContacts()
    }
    
    @objc private func didRecvUserInfoDidChangeNotification(_ notification: Notification) {
        guard let list = notification.userInfo?["userinfo_list"] as? [AgoraChatUserInfo], let contacts = self.contacts else {
            return
        }
        for item in list {
            if let userId = item.userId, contacts.contains(userId) {
                self.table.reloadData()
                break
            }
        }
    }
    
    private func updateContacts(bubbyList: [String]?) {
        self.contacts = bubbyList
        let blackList = AgoraChatClient.shared().contactManager?.getBlackList()
        var contacts: [String] = bubbyList == nil ? [] : Array(bubbyList!)
        if bubbyList != nil, let blackList = blackList {
            for blackId in blackList {
                for i in 0..<contacts.count where contacts[i] == blackId {
                    contacts.remove(at: i)
                    break
                }
            }
        }
        self.sortAndReloadContacts(contacts, subscribe: true)
    }
    
    func reloadContacts() {
        let bubbyList = AgoraChatClient.shared().contactManager?.getContacts()
        self.updateContacts(bubbyList: bubbyList)
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.isSearchState ? 1 : self.sectionTitles.count
    }
    
    func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return self.sectionTitles
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearchState ? self.searchResults.count : self.dataArray[section].count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = table.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDContactCell {
            var model: AgoraUserModel?
            if self.isSearchState {
                model = self.searchResults[indexPath.row] as? AgoraUserModel
            } else {
                model = self.dataArray[indexPath.section][indexPath.row]
            }
            cell.model = model!
            if let hyphenateId = model?.hyphenateId {
                let presence = PresenceManager.shared.presences[hyphenateId]
                let status = PresenceManager.fetchStatus(presence: presence)
                if let imageName = PresenceManager.whiteStrokePresenceImagesMap[status] {
                    cell.iconImageView.presenceImage = UIImage(named: imageName)
                }
                if status != .offline, let statusDescription = presence?.statusDescription, statusDescription.count > 0 {
                    cell.detailLabel.text = statusDescription
                } else {
                    let showStatus = status == .offline ? PresenceManager.formatOfflineTimespace(presence: presence) : PresenceManager.showStatusMap[status]
                    cell.detailLabel.text = showStatus
                }
            }
            cell.tapCellHandle = {
                if let hyphenateId = model?.hyphenateId {
                    self.selectedBlock?(hyphenateId)
                }
            }
        }
        return cell
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
